Implement HappyEyeballsV2AsyncClientConnectionOperator

arturobernalg opened the pull request:
This pull request adds a new implementation of the Happy Eyeballs algorithm for asynchronous client connections in Apache's HttpClient component. The new implementation, HappyEyeballsV2AsyncClientConnectionOperator, builds on the previous version by improving support for IPv6 and providing more fine-grained control over connection timing.

The changes include:
- Addition of HappyEyeballsV2AsyncClientConnectionOperator class and related test classes
- Modifications to existing code to use the new operator where appropriate
- Documentation updates
- Allow activate HappyEyeballsV2Async via Fast Fallback var in PoolingAsyncClientConnectionManagerBuilder.java

ok2c commented:
@arturobernalg Looks good to me. @rschmitt Would you also like to review?

rschmitt commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java:
```diff
@@ -0,0 +1,680 @@
+            if (i < addresses.size() - 1) {
+                try {
+                    final Duration delay = calculateDelay(i);
+                    delay.wait();
```
What thread is this code running in? This will block for up to 2,000 milliseconds.

arturobernalg replied:
The RFC recommends introducing a delay before starting the next connection attempt, but it doesn't mandate a specific implementation. The RFC suggests a default delay of 250 milliseconds, but also notes that a more nuanced implementation can use historical RTT data to influence the delay. Therefore, it is up to the implementation to decide whether to introduce a delay and how to calculate that delay.

rschmitt commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperatorBuilder.java:
```diff
@@ -0,0 +1,272 @@
+
+
+        connectionOperator = Args.notNull(connectionOperator, "Connection operator");
+        final DnsResolver dnsResolverToUse = dnsResolver != null ? dnsResolver : SystemDefaultDnsResolver.INSTANCE;
+        final Timeout timeoutToUse = timeout != null ? timeout : Timeout.ofMilliseconds(250);
+        final Timeout resolutionDelayToUse = resolutionDelay != null ? resolutionDelay : Timeout.ofMilliseconds(50);
+        final Timeout minimumConnectionAttemptDelayToUse = minimumConnectionAttemptDelay != null ? minimumConnectionAttemptDelay : Timeout.ofMilliseconds(100);
+        final Timeout maximumConnectionAttemptDelayToUse = maximumConnectionAttemptDelay != null ? maximumConnectionAttemptDelay : Timeout.ofSeconds(2);
+        final Timeout connectionAttemptDelayToUse = connectionAttemptDelay != null ? connectionAttemptDelay : Timeout.ofMilliseconds(250);
+        final int firstAddressFamilyCountToUse = Args.positive(firstAddressFamilyCount, "First Address Family");
+        final HappyEyeballsV2AsyncClientConnectionOperator.AddressFamily addressFamilyToUse = addressFamily;
```
These defaults are all duplicated in the `HappyEyeballsV2AsyncClientConnectionOperator` constructor.

rschmitt commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/InetAddressComparator.java:
```diff
@@ -0,0 +1,89 @@
+        if (adr1 instanceof Inet4Address && adr2 instanceof Inet6Address) {
+            return -1;
+        }
+        // Compare based on address bytes
```
This appears to have nothing to do with the [address sorting algorithm](http://biplane.com.au/blog/?p=141) specified by [RFC 6724](https://www.rfc-editor.org/rfc/rfc6724#section-6), which is [mandated](https://www.rfc-editor.org/rfc/rfc8305#section-4) by RFC 8305. Furthermore, sorting addresses in this way is a bad idea, since it will defeat [round-robin DNS](https://en.wikipedia.org/wiki/Round-robin_DNS), an important load balancing technique.

rschmitt commented on httpclient5/src/test/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperatorTest.java:
```diff
@@ -0,0 +1,309 @@
+
+    }
+
+    @DisplayName("Test that application prioritizes IPv6 over IPv4 when both are available")
```
Again, the sorting rules are more complicated than this. For example, native IPv4 addresses are preferred to IPv6 addresses that use an encapsulation/tunneling transport like 6to4 or Teredo. IPv6 addresses will also be considered unusable, and therefore deprioritized, if no IPv6 _source_ address is available.

arturobernalg commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java:
```diff
@@ -0,0 +1,680 @@
+            try {
+                final InetAddress[] inetAddresses = dnsResolver.resolve(host);
+                // Introduce a delay before resolving AAAA records after receiving A records
+                resolution_delay.sleep();
```
Instead of using Thread.sleep(), i could use non-blocking methods(ScheduledExecutorService) to delay the execution of the next connection attempt.

rschmitt replied:
Yes, in async code delays should always be implemented by yielding the current thread and scheduling a continuation, rather than calling `Thread.sleep`. Ideally, the DNS lookup would also be made as an async call, but to my knowledge there's really nothing we can do about that, and in practical terms DNS is very fast. I'm not even sure Netty has an async implementation of DNS lookups.

rhernandez35 replied:
Netty does, but it's an entirely bespoke implementation that doesn't touch the JDK's built-int resolver. The blocking here will hopefully become less of a problem for this client once Loom is available.

rhernandez35 commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java:
```diff
@@ -0,0 +1,697 @@
+                    scheduler.schedule(() -> {
+                        try {
+                            final InetAddress[] inet6Addresses = dnsResolver.resolve(host);
+                            dnsFuture.complete(inet6Addresses);
```
You can't complete `dnsFuture` twice. The `complete` call on line 408 will always put `dnsFuture` into a finished state.

rhernandez35 replied:
And please don't use `obtrudeValue` to work around `complete` being one-and-done

arturobernalg commented:
Hi @rschmitt 
just wanted to let you know that I have implemented Destination Address Selection and improved the handling of async connections in HappyEyeballsV2AsyncClientConnectionOperator. 
I have also added a new class to validate the HappyEyeballsV2 rules: HappyEyeballsV2RulesTest. Thought you might find it useful. 

Let me know if you have any feedback. Thanks!

rschmitt commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/InetAddressComparator.java:
```diff
@@ -0,0 +1,470 @@
+                return scope;
+            }
+        }
+        return -1;
```
Isn't `127.0.0.1/8` supposed to have special treatment?

rschmitt commented on httpclient5/src/main/java/org/apache/hc/client5/http/impl/nio/HappyEyeballsV2AsyncClientConnectionOperator.java:
```diff
@@ -0,0 +1,698 @@
+                    }
+
+                    // Sort the array of addresses using the custom Comparator
+                    Arrays.sort(inetAddresses, InetAddressComparator.INSTANCE);
```
Is this a [stable sort](https://www.geeksforgeeks.org/stable-and-unstable-sorting-algorithms/)? It's important to not change the order of IP addresses that are equal under the comparison rules.

rschmitt reviewed (changes requested):
I think a lot of the destination address sorting rules need to be reimplemented in terms of local address availability. It's also important to make sure we are using a stable sort. Finally, the `compareTo` method can't make any network calls.

These rules are complex enough that they really need dedicated unit testing. Generative testing may be a good idea, particularly to ensure that the sort is stable.

arturobernalg commented:
Hi @rschmitt I've made a commit addressing the feedback you provided on the pull request. I believe I've covered all the points. There isn't a specific JUnit test for the rules; there's only an example test class. Currently, I'm looking into how to approach the testing since the class is internal and not exposed at the moment. 
Please feel free to provide any further comments or suggestions.

ok2c commented:
@rschmitt @rhernandez35 Folks,, is there any chance you could do another review sometime soon? @arturobernalg Please in the meantime change the target branch to `5.4.x`.
